crates/extensions/c8y_mapper_ext/src/alarm_converter.rs

```diff
@@ -33,62 +34,67 @@ impl AlarmConverter {
+                    ThinEdgeAlarmDeserializerError::FailedToParsePayloadToString {
                         topic: mqtt_topic.clone(),
                         error: e.to_string(),
                     }
                 })?;
 
-                let tedge_alarm =
-                    ThinEdgeAlarm::try_from(&mqtt_topic, mqtt_payload).map_err(|e| {
-                        ThinEdgeJsonDeserializerError::FailedToParseJsonPayload {
-                            topic: mqtt_topic,
-                            error: e.to_string(),
-                            payload: mqtt_payload.chars().take(50).collect(),
-                        }
-                    })?;
-                let c8y_alarm = C8yCreateAlarm::try_from(&tedge_alarm)?;
+                let tedge_alarm = if mqtt_payload.is_empty() {
+                    // Clearing an alarm
+                    ThinEdgeAlarm::new(alarm_type, source)
+                } else {
+                    // Creating an alarm
+                    let tedge_alarm_data =
+                        ThinEdgeAlarmData::try_from(mqtt_payload).map_err(|e| {
+                            ThinEdgeAlarmDeserializerError::FailedToParseJsonPayload {
+                                topic: mqtt_topic,
+                                error: e.to_string(),
+                                payload: mqtt_payload.chars().take(50).collect(),
+                            }
+                        })?;
+                    ThinEdgeAlarm::new(alarm_type, source).with_alarm_data(tedge_alarm_data)
+                };
+
+                let c8y_alarm = C8yCreateAlarm::from(&tedge_alarm);
 
                 // If the message doesn't contain any fields other than `text` and `time`, convert to SmartREST
-                if c8y_alarm.fragments.is_empty() {
-                    let smartrest_alarm = alarm::serialize_alarm(tedge_alarm)?;
-                    let c8y_alarm_topic = Topic::new_unchecked(
-                        &self.get_c8y_alarm_topic(input_message.topic.name.as_str())?,
-                    );
-                    output_messages.push(Message::new(&c8y_alarm_topic, smartrest_alarm));
-                } else {
-                    let cumulocity_alarm_json = serde_json::to_string(&c8y_alarm)?;
-                    let c8y_alarm_topic = Topic::new_unchecked(C8Y_JSON_MQTT_ALARMS_TOPIC);
-                    output_messages.push(Message::new(&c8y_alarm_topic, cumulocity_alarm_json));
+                match c8y_alarm.data {
+                    Some(ref alarm_data) if !alarm_data.fragments.is_empty() => {
+                        // JSON over MQTT
+                        let cumulocity_alarm_json = serde_json::to_string(&c8y_alarm)?;
+                        let c8y_alarm_topic = Topic::new_unchecked(C8Y_JSON_MQTT_ALARMS_TOPIC);
+                        output_messages.push(Message::new(&c8y_alarm_topic, cumulocity_alarm_json));
+                    }
+                    _ => {
+                        // SmartREST
```

Comment:
Why not sending all the alarms using C8Y JSON?

Reply:
Agree! Still need to check how to clear an alarm by JSON over MQTT though. But I would address this later with https://github.com/thin-edge/thin-edge.io/issues/2242.

Reply:
Just legacy, I guess. But doing that would be much better, in terms of code complexity, as then we just need to create one `C8yCreateAlarm` structure and use that both for MQTT as well as HTTP.